framework/src/eventloop/eventloop_timer.c

```diff
@@ -76,9 +174,27 @@ static el_timer_t *add_timer(el_loop_t *loop, unsigned int timeout, bool repeat,
 		return NULL;
 	}
 	callback->func = func;
-	callback->cb_data = data;
+	if (data_size == 0) {
+		callback->cb_data = NULL;
+	} else {
+		callback->cb_data = EL_ALLOC(data_size);
+		if (callback->cb_data == NULL) {
+			eldbg("Failed to allocate callback data\n");
+			EL_FREE(callback);
+			EL_FREE(timer);
+			return NULL;
+		}
+		memcpy(callback->cb_data, data, data_size);
+	}
 	timer->data = (void *)callback;
 
+	ret = eventloop_register_timer(timer);
+	if (ret != OK) {
+		EL_FREE(callback);
```

Comment:
Is EL_FREE(callback->cb_data) not needed?

Reply:
Done